src/main/java/com/google/devtools/build/lib/rules/repository/RepoRecordedInput.java

```diff
@@ -164,8 +171,97 @@ public abstract boolean isUpToDate(
       Environment env, BlazeDirectories directories, @Nullable String oldValue)
       throws InterruptedException;
 
-  /** Represents a file input accessed during the repo fetch. */
-  public abstract static class File extends RepoRecordedInput {
+  /**
+   * Represents a filesystem path stored in a way that is repo-cache-friendly. That is, if the path
+   * happens to point inside the current Bazel workspace (in either the main repo or an external
+   * repo), we store the appropriate repo name and the path fragment relative to the repo root,
+   * instead of the entire absolute path.
+   *
+   * <p>This is <em>almost</em> like storing a label, but includes the extra corner case of files
+   * inside a repo but not within any package due to missing BUILD files. For example, the file
+   * {@code @@foo//:abc.bzl} is addressable by a label if the file {@code @@foo//:BUILD} exists. But
+   * if the BUILD file doesn't exist, the {@code abc.bzl} file should still be "watchable"; it's
+   * just that {@code @@foo//:abc.bzl} is technically not a valid label.
+   *
+   * <p>Of course, when the path is outside the current Bazel workspace, we just store the absolute
+   * path.
+   */
+  @AutoValue
+  public abstract static class RepoCacheFriendlyPath {
+    public abstract Optional<RepositoryName> repoName();
+
+    public abstract PathFragment path();
+
+    public static RepoCacheFriendlyPath createInsideWorkspace(
+        RepositoryName repoName, PathFragment path) {
+      Preconditions.checkArgument(
+          !path.isAbsolute(), "the provided path should be relative to the repo root");
+      return new AutoValue_RepoRecordedInput_RepoCacheFriendlyPath(Optional.of(repoName), path);
+    }
+
+    public static RepoCacheFriendlyPath createOutsideWorkspace(PathFragment path) {
+      Preconditions.checkArgument(
+          path.isAbsolute(), "the provided path should be absolute in the filesystem");
+      return new AutoValue_RepoRecordedInput_RepoCacheFriendlyPath(Optional.empty(), path);
+    }
+
+    @Override
+    public final String toString() {
+      // We store `@@foo//abc/def:ghi.bzl` as just `@@foo//abc/def/ghi.bzl`. See class javadoc for
+      // more context.
+      return repoName().map(repoName -> repoName + "//" + path()).orElse(path().toString());
+    }
+
+    public static RepoCacheFriendlyPath parse(String s) {
+      if (LabelValidator.isAbsolute(s)) {
+        int doubleSlash = s.indexOf("//");
+        int skipAts = s.startsWith("@@") ? 2 : s.startsWith("@") ? 1 : 0;
+        return createInsideWorkspace(
+            RepositoryName.createUnvalidated(s.substring(skipAts, doubleSlash)),
+            // For backwards compatibility, treat colons as slashes.
+            PathFragment.create(s.substring(doubleSlash + 2).replace(':', '/')));
```

Comment:
FYI: we are planning to allow colons in file names in some cases (cc @c-mita @brandjon ). Mind only replacing the first colon with a slash to be a tiny bit more future-proof?

Reply:
actually i think I'll just remove this replace, there's no 'backwards compatibility' to be had anyway since we're incrementing the marker file version.